test/Interop/Cxx/function/Inputs/function-templates.h

```diff
@@ -0,0 +1,16 @@
+template <class T> T add(T a, T b) { return a + b; }
+
+template <class A, class B> A addTwoTemplates(A a, B b) { return a + b; }
+
+template <class T> T passThrough(T value) { return value; }
+
+template <class T> const T passThroughConst(const T value) { return value; }
+
+// We cannot yet use this in swift but, make sure we don't crash when parsing
+// it.
+template <class R, class T, class U> R returns_template(T a, U b) {
+  return a + b;
+}
+
+// Same here:
+template <class T> void cannot_infer_template() {}
```

Comment:
Can we get some tests for SFINAE? For example showing that the right function out of an overload set is selected

Reply:
Good idea. After writing some SFINAE tests I found an infinite loop. I'll have to fix that before this lands.